dependencydiff/dependencydiff.go

```diff
@@ -0,0 +1,162 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dependencydiff
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/ossf/scorecard/v4/checker"
+	"github.com/ossf/scorecard/v4/checks"
+	"github.com/ossf/scorecard/v4/clients"
+	"github.com/ossf/scorecard/v4/clients/githubrepo"
+	"github.com/ossf/scorecard/v4/log"
+	"github.com/ossf/scorecard/v4/pkg"
+)
+
+// Depdiff is the exported name for dependency-diff.
+const Depdiff = "Dependency-diff"
+
+type dependencydiffContext struct {
+	logger                                *log.Logger
+	ownerName, repoName, baseSHA, headSHA string
+	ctx                                   context.Context
+	ghRepo                                clients.Repo
+	ghRepoClient                          clients.RepoClient
+	ossFuzzClient                         clients.RepoClient
+	vulnsClient                           clients.VulnerabilitiesClient
+	ciiClient                             clients.CIIBestPracticesClient
+	changeTypesToCheck                    map[pkg.ChangeType]bool
+	checkNamesToRun                       []string
+	dependencydiffs                       []dependency
+	results                               []pkg.DependencyCheckResult
+}
+
+// GetDependencyDiffResults gets dependency changes between two given code commits BASE and HEAD
+// along with the Scorecard check results of the dependencies, and returns a slice of DependencyCheckResult.
+// TO use this API, an access token must be set following https://github.com/ossf/scorecard#authentication.
+func GetDependencyDiffResults(
+	ctx context.Context, ownerName, repoName, baseSHA, headSHA string, scorecardChecksNames []string,
+	changeTypesToCheck map[pkg.ChangeType]bool) ([]pkg.DependencyCheckResult, error) {
+	// Fetch the raw dependency diffs.
+	dCtx := dependencydiffContext{
+		logger:             log.NewLogger(log.InfoLevel),
+		ownerName:          ownerName,
+		repoName:           repoName,
+		baseSHA:            baseSHA,
+		headSHA:            headSHA,
+		ctx:                ctx,
+		changeTypesToCheck: changeTypesToCheck,
+		checkNamesToRun:    scorecardChecksNames,
+	}
+	err := fetchRawDependencyDiffData(&dCtx)
+	if err != nil {
+		return nil, fmt.Errorf("error in fetchRawDependencyDiffData: %w", err)
+	}
+
+	// Initialize the repo and client(s) corresponding to the checks to run.
+	err = initRepoAndClientByChecks(&dCtx)
+	if err != nil {
+		return nil, fmt.Errorf("error in initRepoAndClientByChecks: %w", err)
+	}
+	getScorecardCheckResults(&dCtx)
+	return dCtx.results, nil
+}
+
+func initRepoAndClientByChecks(dCtx *dependencydiffContext) error {
```

Comment:
I suspect this function is also in scorecard. Can we call it?

Reply:
Yeah I found `checker/GetClients()` that initalizes the clients, but when it does that, it returns all the clients. I think mine here is better - creating clients depending on which checks to perform :D